// @payos/node/lib/resources/v1/payouts/batch/batch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Batch = void 0;
const resource_1 = require("../../../../core/resource");
class Batch extends resource_1.APIResource {
    /**
     * Create a batch payout.
     *
     * @param {PayoutBatchRequest} payoutData The details of batch payout to be created.
     * @param {string|undefined} idempotencyKey A unique key for ensuring idempotency.
     * @param {RequestOptions<PayoutBatchRequest>} options Additional options.
     * @returns {Promise<Payout>} A promise that resolves to the newly created `Payout` object.
     */
    async create(payoutData, idempotencyKey = this._client.crypto.createUuidv4(), options) {
        const data = await this._client.post('/v1/payouts/batch', {
            signatureOpts: {
                request: 'header',
                response: 'header',
            },
            body: payoutData,
            headers: {
                'x-idempotency-key': idempotencyKey,
            },
            ...options,
        });
        return data;
    }
}
exports.Batch = Batch;
