// @payos/node/lib/resources/v2/payment-requests/payment-requests.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaymentRequests = void 0;
const resource_1 = require("../../../core/resource");
const invoices_1 = require("./invoices");
class PaymentRequests extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.invoices = new invoices_1.Invoices(this._client);
    }
    /**
     * Creates a payment link for the provided order data.
     *
     * @param {CreatePaymentLinkRequest} paymentData The details of payment link to be created.
     * @param {RequestOptions<CreatePaymentLinkRequest>} options Additional options.
     * @returns {Promise<CreatePaymentLinkResponse>}
     *   A promise resolving to a `CreatePaymentLinkResponse` object containing the payment link detail.
     *
     */
    async create(paymentData, options) {
        const data = await this._client.post('/v2/payment-requests', {
            ...options,
            body: paymentData,
            signatureOpts: { request: 'create-payment-link', response: 'body' },
        });
        return data;
    }
    async get(id, options) {
        const data = await this._client.get(`/v2/payment-requests/${id}`, {
            ...options,
            signatureOpts: { response: 'body' },
        });
        return data;
    }
    async cancel(id, cancellationReason, options) {
        const requestData = cancellationReason ? { cancellationReason: cancellationReason } : undefined;
        const data = await this._client.post(`/v2/payment-requests/${id}/cancel`, {
            ...options,
            body: requestData,
            signatureOpts: { response: 'body' },
        });
        return data;
    }
}
exports.PaymentRequests = PaymentRequests;
