// @payos/node/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebhookError = exports.InvalidSignatureError = exports.InternalServerError = exports.TooManyRequestError = exports.NotFoundError = exports.ForbiddenError = exports.UnauthorizedError = exports.BadRequestError = exports.ConnectionTimeoutError = exports.ConnectionError = exports.UserAbortError = exports.APIError = exports.PayOSError = exports.PayOS = void 0;
var client_1 = require("./client");
Object.defineProperty(exports, "PayOS", { enumerable: true, get: function () { return client_1.PayOS; } });
var error_1 = require("./core/error");
Object.defineProperty(exports, "PayOSError", { enumerable: true, get: function () { return error_1.PayOSError; } });
Object.defineProperty(exports, "APIError", { enumerable: true, get: function () { return error_1.APIError; } });
Object.defineProperty(exports, "UserAbortError", { enumerable: true, get: function () { return error_1.UserAbortError; } });
Object.defineProperty(exports, "ConnectionError", { enumerable: true, get: function () { return error_1.ConnectionError; } });
Object.defineProperty(exports, "ConnectionTimeoutError", { enumerable: true, get: function () { return error_1.ConnectionTimeoutError; } });
Object.defineProperty(exports, "BadRequestError", { enumerable: true, get: function () { return error_1.BadRequestError; } });
Object.defineProperty(exports, "UnauthorizedError", { enumerable: true, get: function () { return error_1.UnauthorizedError; } });
Object.defineProperty(exports, "ForbiddenError", { enumerable: true, get: function () { return error_1.ForbiddenError; } });
Object.defineProperty(exports, "NotFoundError", { enumerable: true, get: function () { return error_1.NotFoundError; } });
Object.defineProperty(exports, "TooManyRequestError", { enumerable: true, get: function () { return error_1.TooManyRequestError; } });
Object.defineProperty(exports, "InternalServerError", { enumerable: true, get: function () { return error_1.InternalServerError; } });
Object.defineProperty(exports, "InvalidSignatureError", { enumerable: true, get: function () { return error_1.InvalidSignatureError; } });
Object.defineProperty(exports, "WebhookError", { enumerable: true, get: function () { return error_1.WebhookError; } });
__exportStar(require("./resources"), exports);
